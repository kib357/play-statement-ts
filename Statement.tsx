import React from "react";
import { Invoice, Play } from "./types";

type Props = { invoice: Invoice; plays: Record<string, Play> };

const Statement: React.FC<Props> = ({ invoice, plays }) => {
  const format = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: 2,
  }).format;

  let totalAmount = 0;
  let volumeCredits = 0;

  return (
    <div>
      <h1>Statement for {invoice.customer}</h1>
      <ol>
        {invoice.performances.map((perf, i) => {
          const play = plays[perf.playID];
          let thisAmount = 0;

          switch (play.type) {
            case "tragedy":
              thisAmount = 40000;
              if (perf.audience > 30) {
                thisAmount += 1000 * (perf.audience - 30);
              }
              break;
            case "comedy":
              thisAmount = 30000;
              if (perf.audience > 20) {
                thisAmount += 10000 + 500 * (perf.audience - 20);
              }
              thisAmount += 300 * perf.audience;
              break;
            default:
              throw new Error(`unknown type: ${play.type}`);
          }

          // add volume credits
          volumeCredits += Math.max(perf.audience - 30, 0);
          // add extra credit for every ten comedy attendees
          if ("comedy" === play.type)
            volumeCredits += Math.floor(perf.audience / 5);

          totalAmount += thisAmount;

          return (
            <li key={i}>
              {play.name}: {format(thisAmount / 100)} ({perf.audience} seats)
            </li>
          );
        })}
      </ol>
      <p>Amount owed is {format(totalAmount / 100)}</p>
      <p>You earned {volumeCredits} credits</p>
    </div>
  );
};

export default Statement;
