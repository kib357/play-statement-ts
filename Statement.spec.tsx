import React from "react";
import Statement from "./Statement";
import { create } from "react-test-renderer";
import { Invoice, Play, PlayType } from "./types";

describe("report statement should", () => {
  const invoice: Invoice = {
    customer: "BigCo",
    performances: [
      {
        playID: "hamlet",
        audience: 55,
      },
      {
        playID: "as-like",
        audience: 35,
      },
      {
        playID: "othello",
        audience: 29,
      },
      {
        playID: "merchant",
        audience: 19,
      },
    ],
  };

  test("renders correctly", () => {
    const plays: Record<string, Play> = {
      hamlet: { name: "Hamlet", type: PlayType.tragedy },
      "as-like": { name: "As You Like It", type: PlayType.comedy },
      othello: { name: "Othello", type: PlayType.tragedy },
      merchant: { name: "The Merchant of Venice", type: PlayType.comedy },
    };

    const renderer = create(<Statement plays={plays} invoice={invoice} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
