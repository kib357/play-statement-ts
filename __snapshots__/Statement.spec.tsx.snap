// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report statement should renders correctly 1`] = `
<div>
  <h1>
    Statement for 
    BigCo
  </h1>
  <ol>
    <li>
      Hamlet
      : 
      $650.00
       (
      55
       seats)
    </li>
    <li>
      As You Like It
      : 
      $580.00
       (
      35
       seats)
    </li>
    <li>
      Othello
      : 
      $400.00
       (
      29
       seats)
    </li>
    <li>
      The Merchant of Venice
      : 
      $357.00
       (
      19
       seats)
    </li>
  </ol>
  <p>
    Amount owed is 
    $1,987.00
  </p>
  <p>
    You earned 
    40
     credits
  </p>
</div>
`;
